### PR TITLE
Extend the frr enhancement proposal with the list of tests

### PR DESCRIPTION
**Pull request by fedepaol**
Sending this as discussed with @markdgray . 
The enhancement proposal mentions extending the current e2e test suite.
Here, we provide a list of the coverage needed to ensure compatibility between the old implementation and the new implementation.

This is meant to be a starting point for defining the list of our tests. I think it's a good start, and meant to start a discussion to understand if new scenarios need to be added.

### CODE REVIEW

File: design/0001-frr.md
Comment (markdgray) [inline design/0001-frr.md:348]: We should probably check that keepalives are received?

File: design/0001-frr.md
Comment (markdgray) [inline design/0001-frr.md:347]: Might be good to test publishing of /24s and /32s.

File: design/0001-frr.md
Comment (markdgray) [inline design/0001-frr.md:349]: Should we check non-functional things like failover time or time for changes to propagate? I feel that this may be out of scope but it would be good to get your opinion?

File: design/0001-frr.md
Comment (markdgray) [inline design/0001-frr.md:367]: You mean we need to match the metrics that are currently exposed by the native implementation?

File: design/0001-frr.md
Comment (markdgray) [inline design/0001-frr.md:327]: I guess there are some things we can read from the peer to ensure the connection is as expected. One thing for example is the connection state but maybe there are others? We currently don't do that but it could be useful.

File: design/0001-frr.md
Comment (fedepaol) [reply design/0001-frr.md:348]: Isn't checking the state of the peers / routes enough to verify that the new state is adjusted?

File: design/0001-frr.md
Comment (fedepaol) [reply design/0001-frr.md:327]: Yes, the connection state is already part of my PR https://github.com/metallb/metallb/pull/944 , plus there I check that routes are set as we expect both in FRR and inside the container.

File: design/0001-frr.md
Comment (fedepaol) [reply design/0001-frr.md:367]: Yes, I'll reword that to make it clearer.

File: design/0001-frr.md
Comment (fedepaol) [reply design/0001-frr.md:349]: I am a bit torn about this. 
I'd say, let's start with functional tests assuming that is fair to wait seconds for a config to be in place because we don't have feedback via API to understand if the configuration was applied or not.

File: design/0001-frr.md
Comment (fedepaol) [reply design/0001-frr.md:347]: Will add that

File: design/0001-frr.md
Comment (markdgray) [reply design/0001-frr.md:327]: Yes, I started reviewing your PR after this. Good idea!

File: design/0001-frr.md
Comment (markdgray) [reply design/0001-frr.md:348]: Yes, you are probably right.

**Review by rata — APPROVED** [review]
LGTM, ping me when you finish the changes you wanted to do (https://github.com/metallb/metallb/pull/944/files#r708108762).

IMO, it is enough for the tests be the authoritative source of what is being tested. Such a description can be in the PR description actually adding the tests. But if you prefer to add it to this doc before, I'm fine too :)

**Comment by markdgray** [comment]
Thanks for doing this. Now we have some agreement on what we should be testing and as @rata stated, the tests will be the authoritative source.

**Comment by fedepaol** [comment]
> IMO, it is enough for the tests be the authoritative source of what is being tested. Such a description can be in the PR description actually adding the tests. But if you prefer to add it to this doc before, I'm fine too :)

I agree, this was meant to be both a list of what we need to implement, and a discussion on what we may have missed.

**Comment by fedepaol** [comment]
> LGTM, ping me when you finish the changes you wanted to do (https://github.com/metallb/metallb/pull/944/files#r708108762).
> 
> IMO, it is enough for the tests be the authoritative source of what is being tested. Such a description can be in the PR description actually adding the tests. But if you prefer to add it to this doc before, I'm fine too :)

Finished with the changes

File: design/0001-frr.md
Comment (cgoncalves) [inline design/0001-frr.md:341]: Do we also need to add test coverage for IP address sharing?
https://metallb.universe.tf/usage/#ip-address-sharing

**Review by cgoncalves — COMMENTED** [review]
A great addition to the design document! I particularly like the fact you're calling out the need to test cover many of the existing parameters like communities, localPref, etc.

File: design/0001-frr.md
Comment (fedepaol) [reply design/0001-frr.md:341]: I'd say no, simply because the allocation side of things is not impacted by switching bgp implementation.

Not saying it should not be tested, but here we are aiming at the set of tests that will limit (don't want to say avoid :P ) the regressions between the two implementations.

**Review by russellb — APPROVED** [review]
lgtm, only commented on one very minor typo that I'll fix myself before merging

**Pull request merged**
